zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/controller/DocumentController.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.zeebe.gateway.rest.controller;
+
+import io.camunda.service.DocumentServices;
+import io.camunda.service.DocumentServices.DocumentReferenceResponse;
+import io.camunda.zeebe.gateway.protocol.rest.DocumentLinkRequest;
+import io.camunda.zeebe.gateway.protocol.rest.DocumentMetadata;
+import io.camunda.zeebe.gateway.protocol.rest.DocumentReference;
+import io.camunda.zeebe.gateway.rest.RequestMapper;
+import io.camunda.zeebe.gateway.rest.RestErrorMapper;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.CompletableFuture;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.DeleteMapping;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RequestPart;
+import org.springframework.web.multipart.MultipartFile;
+import org.springframework.web.servlet.mvc.method.annotation.StreamingResponseBody;
+
+@CamundaRestController
+@RequestMapping("/v2/documents")
+public class DocumentController {
+
+  private final DocumentServices documentServices;
+
+  @Autowired
+  public DocumentController(final DocumentServices documentServices) {
+    this.documentServices = documentServices;
+  }
+
+  @PostMapping
+  public CompletableFuture<ResponseEntity<DocumentReference>> createDocument(
```

Comment:
🔧 We consistently return `CompletableFuture<ResponseEntity<Object>>` from controllers because we cannot know if it will be the success type or a `ProblemDetail` in the `ResponseEntity`. Please also have a look at defining the accepted and returned content types of your endpoints in this controller.

Reply:
I did this for upload and delete methods, but couldn't make it work for the get method. The issue is, if we want to return a `StreamingResponseBody`, we need to define it explicitly as a method return type, and not use a generic type, otherwise the response is not handled correctly.
We might work around this by implementing a custom [`HandlerMethodReturnValueHandler`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/method/support/HandlerMethodReturnValueHandler.html), but I think a nicer way is to define an exception handler for this method (see `handleDocumentContentException`), and return specific types from the method and the exception handler instead of returning Object from the method.

We can also think about defining a custom exception for this method to further limit the exception handler scope (currently it would handle methods of this controller) - but other methods are not supposed to throw an unhandled `CamundaServiceException` anyway. WDYT?

Reply:
Hmm, I am not too familiar with the streaming response, to be honest. What does 

>otherwise the response is not handled correctly

explicitly result in? How is it handled when the response is a `ResponseEntity<Object>`? It is not streamed back but provided in one chunk?

AFAIK, the generic is removed by Java at runtime anyway and Spring simply inspects the actual return type/body of the response and handles it accordingly.

I'd need more time to investigate options here, to be honest. I don't have that time currently and I'm on FTO for 2 weeks starting this afternoon 😅 We can also get @nicpuppa and/or @koevskinikola involved next week here to help you with that.

Regarding
> We can also think about defining a custom exception for this method to further limit the exception handler scope (currently it would handle methods of this controller) - but other methods are not supposed to throw an unhandled CamundaServiceException anyway. WDYT?

The exception handler inside the controller is not a good idea, I would say. Currently, no other controller is supposed to simply (re-)throw a `CamundaServiceException`, yes. But that might happen by accident and might result in a weird exception handling then which is not easy to track down. 

If we can't make it easily work with the streaming response, I'd rather use a specific (new) exception here at the moment so we don't introduce side effects in other controllers. You can also add a method for the exception in the `GlobalControllerExceptionHandler`.

Reply:
Maybe using `CompletableFuture<ResponseEntity<Object>>` can help here but we'd need to see how to get the content streaming result in there. I guess it's possible but I really don't know. Just food for thought.

Reply:
> How is it handled when the response is a ResponseEntity Object? It is not streamed back but provided in one chunk?


It serializes the StreamingResponseBody object as if it was a POJO which results in an empty JSON object 😄 
I tried specifying the body content-type of the response entity wrapper as `application/octet-stream` but then it failed with an error (something like `no converter found for type <lambda> and application/octet-strteam`) which made me assume the handler is selected before the method invocation. Wrapping into a CompletableFuture didn't make a difference.

I'll continue the investigation and maybe I'll find an elegant solution, but if not - I also agree with your suggestion to introduce a specific exception for this method 👍

Reply:
I see, OK. Thanks for looking into it, @chillleader 🙏

Reply:
Looks like it actually checks the method return value and inspects the generic type: [StreamingResponseBodyReturnValueHandler](https://github.com/spring-projects/spring-framework/blob/2eff5cb46318eab8e1ed9a7abe27428bc6fb1477/spring-webmvc/src/main/java/org/springframework/web/servlet/mvc/method/annotation/StreamingResponseBodyReturnValueHandler.java#L54)

Reply:
Uh, interesting. Well, if we're going for this result type, a specific exception that is handled accordingly then might be the best way. The global handling should still be generic and only use the information from the thrown exception, if possible.

Reply:
The global exception handler could also simply be improved a bit, for example doing something like the following (just pseudo code, will not work as is in the global handler):

```java
    if (e instanceof IllegalArgumentException) {
      final ProblemDetail problemDetail =
          ProblemDetail.forStatusAndDetail(HttpStatus.BAD_REQUEST, e.getMessage());
      problemDetail.setInstance(URI.create(requestUri));
      return ResponseEntity.of(problemDetail).build();
    }

    final ProblemDetail problemDetail =
        ProblemDetail.forStatusAndDetail(HttpStatus.INTERNAL_SERVER_ERROR, e.getMessage());
    problemDetail.setInstance(URI.create(requestUri));
    return ResponseEntity.of(problemDetail).build();
```

Reply:
It seems we can't modify this behavior unless we set up a custom `ResponseValueHandler`. I added a custom exception and kept the exception handler inside this controller - arguably, it is more readable to have it in the same class as it doesn't impact other controllers. `StreamingResponseBody` seems to be the way to go if we want to return potentially large content in a non-blocking way, so I'd stick to this right now.

Happy to adjust in the future if we find a better alternative

Reply:
Thanks, @chillleader, sounds good, I'll have a look now.